src/main/java/org/opentripplanner/updater/stoptime/TimetableSnapshotSource.java

```diff
@@ -214,30 +215,53 @@ public void applyTripUpdates(final Graph graph, final boolean fullDataset, final
                     // starts for example at 40:00, yesterday would probably be a better guess.
                 }
 
+                if (!tripDescriptor.hasTripId()) {
+                    LOG.warn("No trip id found for gtfs-rt trip update: \n{}", tripUpdate);
+                    continue;
+                }
+
                 uIndex += 1;
                 LOG.debug("trip update #{} ({} updates) :",
                         uIndex, tripUpdate.getStopTimeUpdateCount());
                 LOG.trace("{}", tripUpdate);
 
                 // Determine what kind of trip update this is
                 boolean applied = false;
-                final TripDescriptor.ScheduleRelationship tripScheduleRelationship = determineTripScheduleRelationship(
-                        tripUpdate);
+                final TripDescriptor.ScheduleRelationship tripScheduleRelationship =
+                        determineTripScheduleRelationship(tripDescriptor);
                 switch (tripScheduleRelationship) {
                     case SCHEDULED:
-                        applied = handleScheduledTrip(tripUpdate, feedId, serviceDate);
+                        applied = handleScheduledTrip(
+                                graph,
```

Comment:
I know that you don't want to refactor the world but the tight coupling with the graph makes this class harder and harder to test.

Reply:
https://github.com/opentripplanner/OpenTripPlanner/pull/3960/commits/76ff7db8fc49ab3fa1ae9d9335c409fc27884969 is this to the direction you wanted?